modules/src/ics03_connection/connection.rs

```diff
@@ -0,0 +1,107 @@
+use super::exported::*;
+use crate::ics03_connection::error;
+use crate::ics03_connection::error::Kind;
+use crate::ics23_commitment::CommitmentPrefix;
+use crate::ics24_host::identifier::{ClientId, ConnectionId};
+use serde_derive::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct ConnectionEnd {
+    state: State,
+    client_id: ClientId,
+    counterparty: Counterparty,
+    versions: Vec<String>,
+}
+
+impl ConnectionEnd {
+    pub fn new(client_id: ClientId, counterparty: Counterparty, versions: Vec<String>) -> Self {
+        ConnectionEnd {
+            state: State::Uninitialized,
+            client_id,
+            counterparty,
+            versions,
+        }
+    }
+}
+
+impl ConnectionI for ConnectionEnd {
+    type ValidationError = error::Error;
+
+    fn state(&self) -> State {
+        self.state.clone()
+    }
```

Comment:
It's usually a good practice to not clone data when it can be avoided, and let the call site perform the clone if it needs ownership of the returned data. Not a big deal in this case, as we can incur the penalty of doing a few clones here and there, so feel free to ignore.

```suggestion
    fn state(&self) -> &State {
        &self.state
    }
```